libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
 
     self.acceptFuts[index] = self.servers[index].accept()
 
-    let transp = await finished
-    try:
-      let observedAddr = MultiAddress.init(transp.remoteAddress).tryGet()
-      return await self.connHandler(transp, Opt.some(observedAddr), Direction.In)
-    except CancelledError as exc:
-      debug "CancelledError", exc = exc.msg
-      transp.close()
-      raise exc
-    except CatchableError as exc:
-      debug "Failed to handle connection", exc = exc.msg
-      transp.close()
-  except TransportTooManyError as exc:
-    debug "Too many files opened", exc = exc.msg
-  except TransportAbortedError as exc:
-    debug "Connection aborted", exc = exc.msg
-  except TransportUseClosedError as exc:
-    debug "Server was closed", exc = exc.msg
-    raise newTransportClosedError(exc)
-  except CancelledError as exc:
-    raise exc
-  except TransportOsError as exc:
-    info "OS Error", exc = exc.msg
-    raise exc
-  except CatchableError as exc:
-    info "Unexpected error accepting connection", exc = exc.msg
-    raise exc
+    let remote =
+      try:
+        transp.remoteAddress
+      except TransportOsError as exc:
+        # The connection had errors / was closed before `await` returned control
+        await transp.closeWait()
+        debug "Cannot read remote address", exc = exc.msg
+        return nil
 
-method dial*(
-  self: TcpTransport,
-  hostname: string,
-  address: MultiAddress,
-  peerId: Opt[PeerId] = Opt.none(PeerId)): Future[Connection] {.async.} =
-  ## dial a peer
-  ##
+    let observedAddr =
+      MultiAddress.init(remote).expect("Can initialize from remote address")
+    self.connHandler(transp, Opt.some(observedAddr), Direction.In)
 
-  trace "Dialing remote peer", address = $address
-  let transp =
-    if self.networkReachability == NetworkReachability.NotReachable and self.addrs.len > 0:
-      self.clientFlags.incl(SocketFlags.ReusePort)
-      await connect(address, flags = self.clientFlags, localAddress = Opt.some(self.addrs[0]))
-    else:
-      await connect(address, flags = self.clientFlags)
-
-  try:
-    let observedAddr = MultiAddress.init(transp.remoteAddress).tryGet()
-    return await self.connHandler(transp, Opt.some(observedAddr), Direction.Out)
-  except CatchableError as err:
-    await transp.closeWait()
-    raise err
+  impl(self)
 
-method handles*(t: TcpTransport, address: MultiAddress): bool {.gcsafe.} =
+method dial*(
+    self: TcpTransport,
+    hostname: string,
+    address: MultiAddress,
+    peerId: Opt[PeerId] = Opt.none(PeerId),
+): Future[Connection] =
+  ## dial a peer
+  proc impl(
+      self: TcpTransport, hostname: string, address: MultiAddress, peerId: Opt[PeerId]
+  ): Future[Connection] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if self.stopping:
+      raise newTransportClosedError()
+
+    let ta = initTAddress(address).valueOr:
+      raise (ref TcpTransportError)(msg: "Unsupported address: " & $address)
```

Comment:
why not https://nim-lang.org/docs/system.html#newException.t%2Ctypedesc%2Cstring%2Cref.Exception ?

Reply:
more typing and you can't set custom fields, so `(ref Xxx)` works in more cases when exceptions have rich error information